astropy/cosmology/io/html.py

```diff
@@ -0,0 +1,161 @@
+        ``Cosmology(meta={'key':10}, key=42)``, the ``Cosmology.meta`` will be
+        ``{'key': 10}``).
+    cosmology : str or |Cosmology| class or None, optional keyword-only
+        The cosmology class (or string name thereof) to use when constructing
+        the cosmology instance. The class also provides default parameter
+        values, filling in any non-mandatory arguments missing in 'table'.
+    **kwargs : Any
+        Passed to :attr:`astropy.table.QTable.read`.
+        ``format`` is set to 'ascii.html', regardless of input.
+
+    Returns
+    -------
+    |Cosmology| subclass instance
+    """
+    kwargs["format"] = "ascii.html"  # force the format
+    # Reading is handled by `QTable`.
+    with u.add_enabled_units(cu):  # (cosmology units not turned on by default)
+        table = QTable.read(filename, **kwargs)
+
+    # Need to map the table's column names to Cosmology inputs.
+    table_column_names = set(table.columns)
+    for cosmo_name, html_name in _FORMAT_TABLE.items():
+        if html_name in table_column_names:
+            table.rename_column(html_name, cosmo_name)
+
+    # Build the cosmology from table, using the private backend.
+    return from_table(table, index=index, move_to_meta=move_to_meta, cosmology=cosmology)
+
+
+def write_html_table(cosmology: Cosmology, file, *, overwrite=False, cls=QTable, **kwargs):
+    """Serialize the |Cosmology| into a HTML table.
+
+    Parameters
+    ----------
+    cosmology : |Cosmology| subclass instance file : path-like or file-like
+        Location to save the serialized cosmology.
+    file : path-like or file-like
+        Where to write the html table.
+
+    overwrite : bool, optional keyword-only
+        Whether to overwrite the file, if it exists.
+    cls : |Table| class, optional keyword-only
+        Astropy |Table| (sub)class to use when writing. Default is |QTable|
+        class.
+    **kwargs : Any
+        Passed to ``cls.write``.
+
+    Raises
+    ------
+    TypeError
+        If the optional keyword-argument 'cls' is not a subclass of |Table|.
+
+    Examples
+    --------
+    A HTML file containing a Cosmology HTML table should have scripts enabling
+    MathJax.
+
+    ::
+        <script
+        src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
+        <script type="text/javascript" id="MathJax-script" async
+            src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
+        </script>
+    """
+    # Set cosmology_in_meta as false for now since there is no metadata being kept
+    table = to_table(cosmology, cls=cls, cosmology_in_meta=False)
+
+    cosmo_cls = type(cosmology)
+    n: str
+    for n, col in zip(table.colnames, table.itercols()):
+        param = getattr(cosmo_cls, n, None)
+        if not isinstance(param, Parameter) or param.unit in (None, u.one):
+            continue
+        # Replace column with unitless version
+        table.replace_column(n, (col << param.unit).value, copy=False)

```

Comment:
I don't think it is necessary to get the unitless version, rather you can just change the unit. This might be nicer if the html output gets the option to show units.

Reply:
I don't think that works. The HTML table breaks for columns with units.

```
@pytest.mark.skipif(not HAS_BS4, reason="requires beautifulsoup4")
    def test_to_html_table_bad_index(self, read, write, tmp_path):
        """Test if argument ``index`` is incorrect"""
        fp = tmp_path / "test_to_html_table_bad_index.html"
    
>       write(fp, format="ascii.html")

astropy/cosmology/io/tests/test_html.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
astropy/cosmology/connect.py:121: in __call__
    self.registry.write(self._instance, *args, **kwargs)
astropy/io/registry/core.py:367: in write
    return writer(data, *args, **kwargs)
astropy/cosmology/io/html.py:129: in write_html_table
    table.write(file, overwrite=overwrite, **kwargs)
astropy/table/connect.py:129: in __call__
    self.registry.write(instance, *args, **kwargs)
astropy/io/registry/core.py:367: in write
    return writer(data, *args, **kwargs)
astropy/io/ascii/connect.py:26: in io_write
    return write(table, filename, **kwargs)
astropy/io/ascii/ui.py:849: in write
    lines = writer.write(table)
astropy/io/ascii/html.py:433: in write
    new_col = Column([el[i] for el in col])
astropy/table/column.py:1164: in __new__
    self = super().__new__(
astropy/table/column.py:514: in __new__
    self_data = np.array(data, dtype=dtype, copy=copy)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Quantity 0. eV>

    def __float__(self):
        try:
            return float(self.to_value(dimensionless_unscaled))
        except (UnitsError, TypeError):
>           raise TypeError('only dimensionless scalar quantities can be '
                            'converted to Python scalars')
E           TypeError: only dimensionless scalar quantities can be converted to Python scalars
```

Reply:
Is this a bug in the HTML writer for table?

Reply:
Weird, I tried a simpler version yesterday which did work:
```
qt = QTable([np.arange(3.)*u.m])
qt.write('a.html', format='ascii.html', overwrite=True)
```
However, the html file does not contain the unit, so there is no real benefit. So, best to just leave it for now, I guess.